### PR TITLE
Implement new initializations for conv neuralnet layer

### PR DESCRIPTION
**Pull request by sanuj**
PR for issue #2700 
@lisitsyn Please have a look. I'm not sure if this is the correct way to implement this.

### CODE REVIEW

File: src/shogun/neuralnets/NeuralConvolutionalLayer.cpp [hunk 1]
```diff
@@ -280,6 +292,10 @@ void CNeuralConvolutionalLayer::init()
 	SG_ADD(&m_pooling_height, "pooling_height", "Pooling Height", MS_NOT_AVAILABLE);
 	SG_ADD(&m_stride_x, "stride_x", "X Stride", MS_NOT_AVAILABLE);
 	SG_ADD(&m_stride_y, "stride_y", "Y Stride", MS_NOT_AVAILABLE);
+	
+	// SG_ADD(&m_initialization_mode, "initialization_mode", "Initialization Mode",
```
Comment (sanuj) [inline src/shogun/neuralnets/NeuralConvolutionalLayer.cpp]: Shall I typecast it to `machine_int_t*` like done below or make a new `Parameter::add` and a new entry in `lib/DataType.h`?

File: tests/unit/neuralnets/NeuralNetwork_unittest.cc [hunk 2]
```diff
@@ -225,11 +225,11 @@ TEST(NeuralNetwork, backpropagation_convolutional)
...
 	layers->append_element(new CNeuralInputLayer(6,4));
 	layers->append_element(new CNeuralConvolutionalLayer(
-		CMAF_LOGISTIC,1, 1,1, 1,1, 1,1));
+		CMAF_LOGISTIC, 1, 1, 1, 1, 1, 1, 1));
 	layers->append_element(new CNeuralConvolutionalLayer(
-		CMAF_LOGISTIC,1, 1,1, 1,1, 1,1));
+		CMAF_LOGISTIC, 1, 1, 1, 1, 1, 1, 1));
 	layers->append_element(new CNeuralConvolutionalLayer(
-		CMAF_LOGISTIC,1, 1,1, 1,1, 1,1));
+		CMAF_LOGISTIC, 1, 1, 1, 1, 1, 1, 1, InitializationMode::INIT_AUTO));
```
Comment (sanuj) [inline tests/unit/neuralnets/NeuralNetwork_unittest.cc]: Is this enough or shall i make a new test for `CNeuralConvolutionalLayer`?

File: src/shogun/neuralnets/NeuralConvolutionalLayer.cpp (hunk 1 above)
Comment (lisitsyn) [inline src/shogun/neuralnets/NeuralConvolutionalLayer.cpp]: `machine_int_t*` should be ok

**Comment by sanuj** [comment]
@lisitsyn Made the changes you suggested.

File: src/shogun/neuralnets/NeuralConvolutionalLayer.h [hunk 3]
```diff
@@ -40,6 +40,20 @@
...
+
+/** Specifies the type of initialization of parameters/weights for this layer.
+ */
+enum InitializationMode
+{
+	/** Specify the standard deviation of the Gaussian Distribution at the time
+	 * of initialization of the network */
+	INIT_MANUAL=0,
+	/** Sets standard deviation to
+	 * \sqrt{\frac{2}{input\_width\times input\_height\times input\_channels}}
```
Comment (lisitsyn) [inline src/shogun/neuralnets/NeuralConvolutionalLayer.h]: Please use doxygen syntax for equations like here: https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/kernel/GaussianKernel.h

File: src/shogun/neuralnets/NeuralConvolutionalLayer.h [hunk 4]
```diff
@@ -40,6 +40,20 @@
 
 namespace shogun
 {
+
+/** Specifies the type of initialization of parameters/weights for this layer.
+ */
+enum InitializationMode
```
Comment (lisitsyn) [inline src/shogun/neuralnets/NeuralConvolutionalLayer.h]: EInitializationMode

File: tests/unit/neuralnets/NeuralNetwork_unittest.cc (hunk 2 above)
Comment (lisitsyn) [inline tests/unit/neuralnets/NeuralNetwork_unittest.cc]: Should be enough I think

**Comment by sanuj** [comment]
@lisitsyn sorry for the delay. I had my end-semester exams. made the changes you suggested. Please review.

**Comment by lisitsyn** [comment]
@sanuj this time excuse my delay :) can we name MANUAL thing as NORMAL and AUTO as HE_NORMAL (as per author of the paper, like in Keras https://github.com/fchollet/keras/blob/master/keras/initializations.py) ?

**Comment by sanuj** [comment]
@lisitsyn I have updated the code although for some reason i'm unable to run the unit-tests locally. Whenever I do `make test` after building shogun, I get `make: *** No rule to make target `test'.  Stop.
`

**Comment by sanuj** [comment]
@lisitsyn ping :)

**Comment by lisitsyn** [comment]
Ok this looks ok, thanks!

**Pull request merged**